src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECCurve.cs

```diff
@@ -0,0 +1,430 @@
+                {
+                    throw new CryptographicException(SR.Cryptography_InvalidCurveOid);
+                }
+            }
+            else if (IsExplicit)
+            {
+                bool hasErrors = false;
+
+                if (A == null ||
+                    B == null || B.Length != A.Length ||
+                    G.X == null || G.X.Length != A.Length ||
+                    G.Y == null || G.Y.Length != A.Length ||
+                    Order == null || Order.Length == 0 ||
+                    Cofactor == null || Cofactor.Length == 0)
+                {
+                    hasErrors = true;
+                }
+
+                if (IsPrime)
+                {
+                    if (!hasErrors)
+                    {
+                        if (Prime == null || Prime.Length != A.Length)
+                        {
+                            hasErrors = true;
+                        }
+                    }
+
+                    if (hasErrors)
+                        throw new CryptographicException(SR.Cryptography_InvalidECPrimeCurve);
+                }
+                else if (IsCharacteristic2)
+                {
+                    if (!hasErrors)
+                    {
+                        if (Polynomial == null || Polynomial.Length == 0)
+                        {
+                            hasErrors = true;
+                        }
+                    }
+
+                    if (hasErrors)
+                        throw new CryptographicException(SR.Cryptography_InvalidECCharacteristic2Curve);
+                }
+            }
+            else
+            {
+                // Implicit; if there are any values, throw
+                Debug.Assert(CurveType == ECCurveType.Implicit);
+                if (HasAnyExplicitParameters() || Oid != null)
+                {
+                    throw new CryptographicException(string.Format(SR.Cryptography_InvalidCurve, CurveType.ToString()));
+                }
+            }
+        }
+
+        private bool HasAnyExplicitParameters()
+        {
+            return (A != null ||
+                B != null ||
+                G.X != null ||
+                G.Y != null ||
+                Order != null ||
+                Cofactor != null ||
+                Prime != null ||
+                Polynomial != null ||
+                Seed != null ||
+                Hash != null);
+        }
+
+        /// <summary>
+        /// Represents the type of an elliptic curve.
+        /// </summary>
+        public enum ECCurveType : int
+        {
+            Implicit = 0,
+            PrimeShortWeierstrass = 1,
+            PrimeTwistedEdwards = 2,
+            PrimeMontgomery = 3,
+            Characteristic2 = 4,
+            Named = 5,
+        }
+
+        /// <summary>
+        /// Factory class for creating named curves.
+        /// </summary>
+        public static class NamedCurves
```

Comment:
Just as a stylistic thing, I'd personally put the nested enum and nested accelerator classes in separate partial files.